File: app/src/main/java/com/example/android/customviews/charting/DrawerView.java
package com.example.android.customviews.charting;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Scroller;
import android.widget.Toast;

import com.example.android.customviews.R;

import java.util.ArrayList;
import java.util.List;

public class DrawerView extends ViewGroup {

    private static final String TAG = "DrawerView";
    private static final int DEFAULT_CHILD_WIDTH = 128;
    private boolean isOpen;
    private Context mContext;
    private float contentWidth;

    private List<BottomView> mChildren = new ArrayList<BottomView>();
    private TopView mForegroundView;

    public DrawerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initAttribute(context, attrs);
    }

    public DrawerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initAttribute(context, attrs);
    }

    public DrawerView(Context context) {
        super(context);
        init(context);
    }

    private void init(Context context) {
        setWillNotDraw(false);
        mContext = context;
        mForegroundView = new TopView(context);
        mForegroundView.setDrawColor(Color.CYAN);
        addView(mForegroundView);

        mForegroundView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(mContext, "foreground view", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void initAttribute(Context context, AttributeSet attrs) {
        TypedArray arr = context.getTheme().obtainStyledAttributes(attrs,
                R.styleable.DrawerView, 0, 0);

        try {
            isOpen = arr.getBoolean(R.styleable.DrawerView_open, false);

        } finally {
            arr.recycle();
        }
        init(context);
    }

    public void addItem(String text, OnClickListener l) {
        // FIXME 设置item颜色

        BottomView view = new BottomView(mContext);
        view.setDrawColor(Color.BLACK);
        if (mChildren.size() > 0)
            view.setDrawColor(Color.GREEN);
        mChildren.add(view);
        addView(view);
        mForegroundView.setMaxOpenLength(mChildren.size() * DEFAULT_CHILD_WIDTH);
        view.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
//                Log.v(TAG, "on click view:" + v);
                Toast.makeText(mContext, "view:" + v, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private boolean isOpen() {
        int childLength = mChildren.size() * DEFAULT_CHILD_WIDTH;
        int sx = mForegroundView.getScrollX();
        return childLength == sx;
    }

    private boolean isTouchOnChildren(float pointx, int childIndex) {
        int left = (int) (contentWidth - DEFAULT_CHILD_WIDTH * mChildren.size() + DEFAULT_CHILD_WIDTH * childIndex);
        int right = left + DEFAULT_CHILD_WIDTH;
        return pointx >= left && pointx < right;
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {

        if (isOpen()) {
            //FIXME 确认调用子类的dispatchTouchEvent是不是正确的选择
            float x = ev.getX();
            if (x >= 0 && x < contentWidth - DEFAULT_CHILD_WIDTH * mChildren.size()) {
                //落在前景上
                return mForegroundView.dispatchTouchEvent(ev);
            } else {
                for (int i = 0; i < mChildren.size(); i++) {
                    if (isTouchOnChildren(x, i)) {
                        return mChildren.get(i).dispatchTouchEvent(ev);
                    }
                }
            }
        } else {
            return mForegroundView.dispatchTouchEvent(ev);
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        Log.d(TAG, "padding left:" + getPaddingLeft() + "  top:"
                + getPaddingTop() + "  right:" + getPaddingRight()
                + "  bottom:" + getPaddingBottom());
        Log.v(TAG, "new width:" + w + "  new height:" + h + "  old width:"
                + oldw + "  old height:" + oldh);

        float xpadding = (float) (getPaddingLeft() + getPaddingRight());
        float ypadding = (float) (getPaddingTop() + getPaddingBottom());

        contentWidth = (float) w - xpadding;
        int contentHeight = (int) (h - ypadding);

        mForegroundView.layout(getPaddingLeft(), getPaddingTop(), (int) contentWidth, contentHeight);
        mForegroundView.bringToFront();

        int childrenSize = mChildren.size();
        for (int i = 0; i < childrenSize; i++) {
            BottomView view = mChildren.get(i);
            int r = (int) (contentWidth - i * DEFAULT_CHILD_WIDTH);
            int l = r - DEFAULT_CHILD_WIDTH;
            view.layout(l, 0, r, contentHeight);
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
    }

    private static class BottomView extends View {

        private Paint mPaint;
        private Rect mDrawRect;

        public BottomView(Context context) {
            super(context);
            init(context);
        }

        public BottomView(Context context, AttributeSet attrs) {
            super(context, attrs);
            init(context);
        }

        public BottomView(Context context, AttributeSet attrs, int defStyleAttr) {
            super(context, attrs, defStyleAttr);
            init(context);
        }

        private void init(Context context) {
            setupPaint();

            mDrawRect = new Rect(0, 0, DEFAULT_CHILD_WIDTH, DEFAULT_CHILD_WIDTH);
        }

        private void setupPaint() {
            mPaint = new Paint();
            mPaint.setColor(Color.BLACK);
            mPaint.setStyle(Paint.Style.FILL);
        }

        public void setDrawColor(int color) {
            mPaint.setColor(color);
        }

        @Override
        public boolean onTouchEvent(MotionEvent event) {
            return super.onTouchEvent(event);
        }

        @Override
        public void layout(int l, int t, int r, int b) {
            setBackgroundSize(l,t,r,b);
            super.layout(l, t, r, b);
        }

        private void setBackgroundSize(int l, int t, int r, int b) {
            int w = r - l;
            int h = b - t;
            mDrawRect.set(0, 0, w, h);
        }

        @Override
        protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);
            canvas.drawRect(mDrawRect, mPaint);
        }
    }

    private static class TopView extends View {

        private Paint mBackgroundPaint;
        private Rect mBackgroundRect;
        private Scroller mScroller;
        private float lastx;
        private float startx;
        private float dragableDistance = 144.0f;
        private int maxOpenLength;
        private boolean isTouchIn;

        public TopView(Context context, AttributeSet attrs, int defStyle) {
            super(context, attrs, defStyle);
            init(context);
        }

        public TopView(Context context, AttributeSet attrs) {
            super(context, attrs);
            init(context);
        }

        public TopView(Context context) {
            super(context);
            init(context);
        }

        public void setDrawColor(int color) {
            mBackgroundPaint.setColor(color);
        }

        public void setMaxOpenLength(int maxOpenLength) {
            this.maxOpenLength = maxOpenLength;
        }

        private void init(Context context) {
            setupPaint();
            mBackgroundRect = new Rect(0, 0, DEFAULT_CHILD_WIDTH, DEFAULT_CHILD_WIDTH);
            mScroller = new Scroller(context);
        }

        private void setupPaint() {
            mBackgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            mBackgroundPaint.setStyle(Paint.Style.FILL);
        }

        @Override
        public void computeScroll() {
            smoothScrollToNewPlace();
        }

        private void smoothScrollToNewPlace() {
            if (mScroller.computeScrollOffset()) {
                scrollTo(mScroller.getCurrX(), mScroller.getCurrY());
                invalidate();
            }
        }

        private boolean isInDragableDistance(float currentScrollDistance) {
            return -dragableDistance <= currentScrollDistance && currentScrollDistance <= 2 * dragableDistance;
        }

        private boolean isCloseToOpen(float currentScrollDistance) {
            return currentScrollDistance > DEFAULT_CHILD_WIDTH;
        }

        @Override
        public boolean onTouchEvent(MotionEvent event) {
            float currentx;
            float deltax;
            float currentScrollDistance;
            currentx = event.getX();
            deltax = lastx - currentx;
            currentScrollDistance = getScrollX() + deltax;
            lastx = currentx;
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    isTouchIn = isTouchIn(currentx);
                    if (isTouchIn) {
                        startx = event.getX();
                        lastx = startx;
                        return true;

                    }
                    break;
                case MotionEvent.ACTION_MOVE:
//FIXME onClickListener被屏蔽
                    Log.i(TAG, "delta x:" + deltax);
                    if (isInDragableDistance(currentScrollDistance)) {
                        scrollBy((int) deltax, getScrollY());
                    }
                    return true;

                case MotionEvent.ACTION_UP:

                    if (isCloseToOpen(currentScrollDistance)) {
                        scrollToOpenPosition();
                    } else {
                        scrollToClosePosition();
                    }
                    isTouchIn = false;
                    return true;
            }

            return super.onTouchEvent(event);
        }

        private boolean isTouchIn(float touchDownX) {
            int right = getRight() - getScrollX();
            return touchDownX <= right;
        }

        private void scrollToOpenPosition() {
            int distance = maxOpenLength - getScrollX();
            mScroller.startScroll(getScrollX(), 0, distance, 0);
            postInvalidate();
        }

        private void scrollToClosePosition() {
            mScroller.startScroll(getScrollX(), 0, -getScrollX(), 0);
            postInvalidate();
        }

        @Override
        protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);
            canvas.drawRect(mBackgroundRect, mBackgroundPaint);
        }

        @Override
        public void layout(int l, int t, int r, int b) {
            setBackgroundRectSize(l,t,r,b);

            super.layout(l, t, r, b);
        }

        private void setBackgroundRectSize(int l, int t, int r, int b) {
            int w = r - l;
            int h = b - t;
            mBackgroundRect.set(0, 0, w, h);
        }
    }
}

File: app/src/main/java/com/example/android/customviews/drawer/DrawerView.kt
package com.example.android.customviews.drawer

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.Scroller
import android.widget.Toast
import com.ggdsn.slidingview.R
import java.util.*

@Deprecated("用SlidingView")
class DrawerView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : ViewGroup(context, attrs, defStyleAttr) {
    private var isOpen: Boolean = false
    private var contentWidth: Float = 0.toFloat()
    private val children = ArrayList<BottomView>()
    private var foregroundView: TopView

    init {
        if (attrs != null) {
            val arr = context.theme.obtainStyledAttributes(attrs,
                    R.styleable.DrawerView, 0, 0)
            try {
                isOpen = arr.getBoolean(R.styleable.DrawerView_open, false)
            } finally {
                arr.recycle()
            }
        }
        setWillNotDraw(false)
        val foregroundView = TopView(context)
        foregroundView.setDrawColor(Color.CYAN)
        foregroundView.setOnClickListener {
            Toast.makeText(context, "foreground view", Toast.LENGTH_SHORT).show()
        }
        addView(foregroundView)
        this.foregroundView = foregroundView
    }

    fun addItem(text: String, l: View.OnClickListener? = null) {
        // FIXME 设置item颜色
        val view = BottomView(context)
        view.setDrawColor(Color.BLACK)
        if (children.size > 0)
            view.setDrawColor(Color.GREEN)
        children.add(view)
        addView(view)
        foregroundView.setMaxOpenLength(children.size * DEFAULT_CHILD_WIDTH)
        view.setOnClickListener { v ->
            Toast.makeText(context, "view:" + v, Toast.LENGTH_SHORT).show()
        }
    }

    private fun isOpen(): Boolean {
        val childLength = children.size * DEFAULT_CHILD_WIDTH
        val sx = foregroundView.scrollX
        return childLength == sx
    }

    private fun isTouchOnChildren(pointx: Float, childIndex: Int): Boolean {
        val left = (contentWidth - DEFAULT_CHILD_WIDTH * children.size + DEFAULT_CHILD_WIDTH * childIndex).toInt()
        val right = left + DEFAULT_CHILD_WIDTH
        return pointx >= left && pointx < right
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        if (isOpen()) {
            /** FIXME 确认调用子类的dispatchTouchEvent是不是正确的选择
             * 2017.11.19 应该不是，父类应该处理滑动逻辑，如果不是滑动逻辑，再交给子类处理，否则不能轻易地替换子view
             */
            val x = ev.x
            if (x >= 0 && x < contentWidth - DEFAULT_CHILD_WIDTH * children.size) {
                //落在前景上
                return foregroundView.dispatchTouchEvent(ev)
            } else {
                for (i in children.indices) {
                    if (isTouchOnChildren(x, i)) {
                        return children[i].dispatchTouchEvent(ev)
                    }
                }
            }
        } else {
            return foregroundView.dispatchTouchEvent(ev)
        }
        return super.dispatchTouchEvent(ev)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        Log.d(TAG, "padding left:" + paddingLeft + "  top:"
                + paddingTop + "  right:" + paddingRight
                + "  bottom:" + paddingBottom)
        Log.v(TAG, "new width:" + w + "  new height:" + h + "  old width:"
                + oldw + "  old height:" + oldh)

        val xpadding = (paddingLeft + paddingRight).toFloat()
        val ypadding = (paddingTop + paddingBottom).toFloat()

        contentWidth = w.toFloat() - xpadding
        val contentHeight = (h - ypadding).toInt()

        foregroundView.layout(paddingLeft, paddingTop, contentWidth.toInt(), contentHeight)
        foregroundView.bringToFront()

        val childrenSize = children.size
        for (i in 0 until childrenSize) {
            val view = children[i]
            val r = (contentWidth - i * DEFAULT_CHILD_WIDTH).toInt()
            val l = r - DEFAULT_CHILD_WIDTH
            view.layout(l, 0, r, contentHeight)
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
    }

    private class BottomView @JvmOverloads constructor(
            context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
    ) : View(context, attrs, defStyleAttr) {

        private var paint: Paint = Paint().apply {
            color = Color.BLACK
            style = Paint.Style.FILL
        }
        private var mDrawRect: Rect = Rect(0, 0, DEFAULT_CHILD_WIDTH, DEFAULT_CHILD_WIDTH)

        fun setDrawColor(color: Int) {
            paint.color = color
        }

        override fun layout(l: Int, t: Int, r: Int, b: Int) {
            setBackgroundSize(l, t, r, b)
            super.layout(l, t, r, b)
        }

        private fun setBackgroundSize(l: Int, t: Int, r: Int, b: Int) {
            val w = r - l
            val h = b - t
            mDrawRect.set(0, 0, w, h)
        }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            canvas.drawRect(mDrawRect, paint)
        }
    }

    private class TopView @JvmOverloads constructor(
            context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
    ) : View(context, attrs, defStyleAttr) {

        private var mBackgroundPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.FILL
        }
        private var mBackgroundRect: Rect = Rect(0, 0, DEFAULT_CHILD_WIDTH, DEFAULT_CHILD_WIDTH)
        private var mScroller: Scroller = Scroller(context)
        private var lastx: Float = 0.toFloat()
        private var startx: Float = 0.toFloat()
        private val dragableDistance = 144.0f
        private var maxOpenLength: Int = 0
        private var isTouchIn: Boolean = false

        fun setDrawColor(color: Int) {
            mBackgroundPaint.color = color
        }

        fun setMaxOpenLength(maxOpenLength: Int) {
            this.maxOpenLength = maxOpenLength
        }

        override fun computeScroll() {
            smoothScrollToNewPlace()
        }

        private fun smoothScrollToNewPlace() {
            if (mScroller.computeScrollOffset()) {
                scrollTo(mScroller.currX, mScroller.currY)
                invalidate()
            }
        }

        private fun inDraggableDistance(currentScrollDistance: Float): Boolean =
                -dragableDistance <= currentScrollDistance && currentScrollDistance <= 2 * dragableDistance

        private fun isNearOpen(currentScrollDistance: Float): Boolean = currentScrollDistance > DEFAULT_CHILD_WIDTH

        override fun onTouchEvent(event: MotionEvent): Boolean {
            val x: Float
            val deltaX: Float
            val offset: Float
            x = event.x
            deltaX = lastx - x
            offset = scrollX + deltaX
            lastx = x
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    isTouchIn = isTouchIn(x)
                    if (isTouchIn) {
                        startx = event.x
                        lastx = startx
                        return true

                    }
                }
                MotionEvent.ACTION_MOVE -> {
                    //FIXME onClickListener被屏蔽
                    Log.i(TAG, "delta x:" + deltaX)
                    if (inDraggableDistance(offset)) {
                        scrollBy(deltaX.toInt(), scrollY)
                    }
                    return true
                }

                MotionEvent.ACTION_UP -> {
                    if (isNearOpen(offset)) {
                        scrollToOpenPosition()
                    } else {
                        scrollToClosePosition()
                    }
                    isTouchIn = false
                    return true
                }
            }
            return super.onTouchEvent(event)
        }

        private fun isTouchIn(touchDownX: Float): Boolean {
            val right = right - scrollX
            return touchDownX <= right
        }

        private fun scrollToOpenPosition() {
            val distance = maxOpenLength - scrollX
            mScroller.startScroll(scrollX, 0, distance, 0)
            postInvalidate()
        }

        private fun scrollToClosePosition() {
            mScroller.startScroll(scrollX, 0, -scrollX, 0)
            postInvalidate()
        }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            canvas.drawRect(mBackgroundRect, mBackgroundPaint)
        }

        override fun layout(l: Int, t: Int, r: Int, b: Int) {
            setBackgroundRectSize(l, t, r, b)

            super.layout(l, t, r, b)
        }

        private fun setBackgroundRectSize(l: Int, t: Int, r: Int, b: Int) {
            val w = r - l
            val h = b - t
            mBackgroundRect.set(0, 0, w, h)
        }
    }

    companion object {

        private val TAG = "DrawerView"
        private val DEFAULT_CHILD_WIDTH = 128
    }
}
